### ISSUE TITLE
Repeat tests if failed in CI

### ISSUE THREAD

**Issue by capital-G** [supercollider/supercollider#6362]
## Motivation

Due to the containerization, lack of audio hardware and realtime/priority access, the tests in the CI can be flaky and often require a rerun - this creates unnecessary noise in the dev process.

## Description of Proposed Feature

As the actual success rate has a stochastic component due to the environment stated above, it would be possible to be more sure if a test actually fails or just *sometimes* fails by repeating it a set number of times, e.g. n=3.
It could be beneficial if a "cooldown" period in between of e.g. 1 seconds, because it was suggested that the CPU usage of the host machine could be a cause for the problems, see https://github.com/supercollider/supercollider/issues/5360#issuecomment-1601615272 .
Maybe it would be even better to "rerun" the tests at the end of a run as the CPU spike could be vanished by then?

## Plan for Implementation

I had a look into this, but frankly I don't understand testing and exception handling in sclang and the testsuite code seems to be quite nested which is hard to follow what actually happens.
I tried the following but it didn't worked out? It failed or ran into a seg faulted, dependent on which tests were run. But also no "debug printing" occured, so I assume the tests were not even run?

```diff
diff --git a/testsuite/scripts/sclang_test_runner.scd b/testsuite/scripts/sclang_test_runner.scd
index 611c4cddc7..a76540be08 100644
--- a/testsuite/scripts/sclang_test_runner.scd
+++ b/testsuite/scripts/sclang_test_runner.scd
@@ -1,10 +1,19 @@
-// this script is meant to be run with two arguments: path to test prototype indicating tests to be run, and path to test result
+/*
+this script is meant to be run with three arguments:
+1) path to test prototype indicating tests to be run,
+2) path to test result, and
+3) number of retries on a failed test - this is used within CI
+*/
+
 (
 ~exitWhenDone = false;
+~testRetries = 0;
+
 if(thisProcess.argv.size > 0) {
 	"passed args: %".format(thisProcess.argv).postln;
 	~testProto = thisProcess.argv[0];
 	~testResult = thisProcess.argv[1];
+	~testRetries = thisProcess.argv[2] ? 0;
 	~exitWhenDone = true;
 } {
 	~testProto = PathName(thisProcess.nowExecutingPath).pathOnly +/+ "test_run_proto_gha.scd";
@@ -18,6 +27,7 @@ if(File.exists(~testProto).not) {
 
 "~testProto path: %".format(~testProto).postln;
 "~testResult path: %".format(~testResult).postln;
+"~testRetries: %".format(~testRetries).postln;
 
 ~verboseSetup = false; // true or false
 
@@ -63,7 +73,7 @@ if(File.exists(~testProto).not) {
 	testsDict;
 };
 
-~testrun = {|protoFile, resultFile|
+~testrun = {|protoFile, resultFile, testRetries|
 	var tests = ~getTestRecord.(protoFile);
 	var skipped, toExpand, toExclude;
 	var currentSuite = "";
@@ -188,10 +198,27 @@ if(File.exists(~testProto).not) {
 						~writeResult.(tests, resultFile);
 						startTime = Date.localtime();
 
-						try {
-							result = class.new.runTestMethod(script);
-						} { |error|
-							class.new.failed(script, error.errorString, true, nil);
+						// block is our break in a do while block
+						test[\retries] = 0;
+						block {|break|
+							var err;
+							(0..testRetries).do({|numRetry|
+								try {
+									"A test will be executed".postln;
+									result = class.new.runTestMethod(script);
+									break.();
+								} { |error|
+									err = error;
+									test[\retries] = numRetry+1;
+									"Test failed - retry run".postln;
+								};
+							});
+							class.new.failed(
+								method: script,
+								message: err.errorString,
+								report: true,
+								details: nil
+							);
 						};
 
 						endTime = Date.localtime();
@@ -233,7 +260,7 @@ if(File.exists(~testProto).not) {
 
 
 {
-	~testrun.(~testProto, ~testResult);
+	~testrun.(~testProto, ~testResult, ~testRetries);
 	"\n\n******** DONE ********\n\n".postln;
 	// 0.exit;
 	if(~exitWhenDone) {

```

<details>

```
❯ ./sclang ~/github/supercollider/testsuite/scripts/sclang_test_runner.scd ~/github/supercollider/testsuite/scripts/test_run_proto_gha.scd hello2.test 4
compiling class library...
        Found 859 primitives.
        Compiling directory '/Users/scheiba/github/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/SCClassLibrary'
        Compiling directory '/Library/Application Support/SuperCollider/Extensions'
        Compiling directory '/Users/scheiba/Library/Application Support/SuperCollider/Extensions'
        Compiling directory '/Users/scheiba/Library/Application Support/SuperCollider/downloaded-quarks/SafetyNet'
        Compiling directory '/Users/scheiba/Library/Application Support/SuperCollider/downloaded-quarks/JITLibExtensions'
        Compiling directory '/Users/scheiba/Library/Application Support/SuperCollider/downloaded-quarks/AudacityLabels'
        numentries = 1529188 / 25060144 = 0.061
        6499 method selectors, 3856 classes
        method table size 25391856 bytes, big table size 200481152
        Number of Symbols 17023
        Byte Code Size 516462
        compiled 723 files in 1.13 seconds

Info: 11 methods are currently overwritten by extensions. To see which, execute:
MethodOverride.printAll

compile done
localhost : setting clientID to 0.
internal : setting clientID to 0.
Safety('internal') enabled.
Safety('localhost') enabled.
Class tree inited in 0.01 seconds
WARNING: Extension in '/Users/scheiba/Library/Application Support/SuperCollider/Extensions/DS/WindowNoDuplicates.sc' overwrites Meta_Window:new in main class library.
Intentional overwrites must be put in a 'SystemOverwrites' subfolder.


*** Welcome to SuperCollider 3.14.0-dev. *** For help type cmd-d.
passed args: [/Users/scheiba/github/supercollider/testsuite/scripts/test_run_proto_gha.scd, hello2.test, 4]
~testProto path: /Users/scheiba/github/supercollider/testsuite/scripts/test_run_proto_gha.scd
~testResult path: hello2.test
~testRetries: 4


        *** Running the tests ***



        Suite: TestCondition

        Test: *


        Suite: TestCondVar

        Test: test_waitFor_withPredicate_signaledButFalseSoTimedOut


        Suite: TestCoreUGens

        Test: test_ugen_generator_equivalences


        Suite: TestDocument

        Test: *


        Suite: TestDocumentEnvironment

        Test: *


        Suite: TestEnvironmentDocument

        Test: *


        Suite: TestEnvGen_server

        Test: test_zero_gate_n_off_not_sent

        Test: test_trig_gate_nonsustain_env


        Suite: TestLinkClock

        Test: *


        Suite: TestNodeProxy

        Test: test_copy_nodeMapIsCopied


        Suite: TestNodeProxyRoles

        Test: test_proxyroles_xfade_smoothly


        Suite: TestNodeProxy_Server

        Test: test_schedAfterFade_notBeforeQuant


        Suite: TestOSCFunc

        Test: test_argSizeMatching_Succeeds


        Suite: TestSCDoc

        Test: test_helpSourceDirs_includedExtensions


        Suite: TestTempoClock

        Test: test_nextTimeOnGrid_okAfterMeterChange


        Suite: TestThreadReschedule

        Test: test_Threads_rescheduleToOtherClock


        Suite: TestUGen_RTAlloc

        Test: test_FFT


        Suite: TestVolume

        Test: test_booting


        Suite: TestWebView

        Test: test_findText_found

        Test: test_findText_notFound


        Suite: TestMIDI

        Test: test_linuxRealtimeCodesAreCorrect


        Suite: TestMovingSum

        Test: test_runningSums
ERROR: Number-forSeries : first, second or last not an Integer or Float.
ERROR: Primitive '_IdentDict_Put' failed.
Failed.
RECEIVER:
   Integer 0
PATH: /Users/scheiba/github/supercollider/testsuite/scripts/sclang_test_runner.scd

PROTECTED CALL STACK:
        Meta_MethodError:new    0x7fe970242180
                arg this = PrimitiveFailedError
                arg what = Failed.
                arg receiver = 0
        Meta_PrimitiveFailedError:new   0x7fe970249fc0
                arg this = PrimitiveFailedError
                arg receiver = 0
        Object:primitiveFailed  0x7fe9981fab80
                arg this = 0
        Number:forSeries        0x7fe998f0ff40
                arg this = 0
                arg second = nil
                arg last = 4
                arg function = a Function
                var step = 0.0
                var j = 0
        a FunctionDef   0x7fe9905d89d8
                sourceCode = "<an open Function>"
                arg break = a Function
                var err = nil
        a FunctionDef   0x7fe9905aeb78
                sourceCode = "<an open Function>"
                arg test = Dictionary[(test -> test_runningSums), (retries -> 0), (completed -> false), (suite -> TestMovingSum)]
                var class = TestMovingSum
                var testname = test_runningSums
                var script = TestMovingSum:test_runningSums
                var result = nil
                var oldFailures = IdentitySet[]
                var oldPasses = IdentitySet[]
                var newPasses = nil
                var newFailures = nil
                var startTime = Tue Jun 18 11:27:59 2024
                var endTime = nil
                var success = false
        ArrayedCollection:do    0x7fe999095f40
                arg this = [Dictionary[(test -> *), (skipReason -> UnitTest fails when run in GHA (FIXME)), (completed -> true), (skip -> true), (suite -> TestCondition), 
  (error -> Class nil not found)], Dictionary[(test -> test_waitFor_withPredicate_signaledButFalseSoTimedOut), (skipReason -> UnitTest fails when run in GHA (FIXME)), (completed -> true), (skip -> true), (suite -> TestCondVar)], Dictionary[(test -> test_ugen_generator_equivalences), (skipReason -> UnitTest has bugs (FIXME)), (completed -> true), (skip -> true), (su...etc...
                arg function = a Function
                var i = 21
        a FunctionDef   0x7fe9902c4868
                sourceCode = "{
        ~testrun.(~testProto, ~testResult, ~testRetries);
        \"\\n\\n******** DONE ********\\n\\n\".postln;
        // 0.exit;
        if(~exitWhenDone) {
                0.exit;
        }; // exit only if the script was called with arguments
}"
        Routine:prStart 0x7fe998c77c40
                arg this = a Routine
                arg inval = 1.414098944

CALL STACK:
        MethodError:reportError
                arg this = <instance of PrimitiveFailedError>
        < closed FunctionDef >
                arg error = <instance of PrimitiveFailedError>
        Integer:forBy
                arg this = 0
                arg endval = 0
                arg stepval = 2
                arg function = <instance of Function>
                var i = 0
                var j = 0
        SequenceableCollection:pairsDo
                arg this = [*2]
                arg function = <instance of Function>
        Scheduler:seconds_
                arg this = <instance of Scheduler>
                arg newSeconds = 1.431112886
        Meta_AppClock:tick
                arg this = <instance of Meta_AppClock>
                var saveClock = <instance of Meta_SystemClock>
        Process:tick
                arg this = <instance of Main>
^^ The preceding error dump is for ERROR: Primitive '_IdentDict_Put' failed.
Failed.
RECEIVER: 0


^C
```

</details>

Maybe someone with more proficient skills in sclang could give this a go or hive a hint? :) It would be greatly appreciated.

**Comment by smoge** [comment]
The last time I checked this out, I thought there was a new scaling process for each test. 

Can you confirm this? If that's the case, we should keep in mind that Sclang is NOT optimized for startup time. That's why it does not save bytecode files of the class lib to reuse next time since Sclang is not designed to do the same things as Python, for example.

If testing time is a concern, one possibility is to keep testing in the same Sclang process and, in case something goes wrong, restart tests from that point in a new Sclang process.

Did I make sense?

**Comment by capital-G** [comment]
Another try. I think it isn't really hard, but I really don't understand the test runner script...

Do you wanna chime in here @dyfer ? I think this could really help us in everyday life and reduce CI waiting time^^

```diff
diff --git a/testsuite/scripts/sclang_test_runner.scd b/testsuite/scripts/sclang_test_runner.scd
index 611c4cddc7..4ca57189f5 100644
--- a/testsuite/scripts/sclang_test_runner.scd
+++ b/testsuite/scripts/sclang_test_runner.scd
@@ -1,6 +1,7 @@
 // this script is meant to be run with two arguments: path to test prototype indicating tests to be run, and path to test result
 (
 ~exitWhenDone = false;
+~numRetries = 3;
 if(thisProcess.argv.size > 0) {
 	"passed args: %".format(thisProcess.argv).postln;
 	~testProto = thisProcess.argv[0];
@@ -179,6 +180,8 @@ if(File.exists(~testProto).not) {
 
 					script = class.findTestMethods().detect({ |m| m.name == testname });
 					if (script.notNil) {
+						var testSuccess = false;
+						var testRetry = 0;
 						// This is a bad way to get the results, but there's no other for now.
 						// One this is working end-to-end, then UnitTest can be improved incrementally.
 						oldPasses = IdentitySet.newFrom(class.passes);
@@ -188,11 +191,24 @@ if(File.exists(~testProto).not) {
 						~writeResult.(tests, resultFile);
 						startTime = Date.localtime();
 
-						try {
-							result = class.new.runTestMethod(script);
-						} { |error|
-							class.new.failed(script, error.errorString, true, nil);
-						};
+						while({testSuccess.not.and(testRetry < ~numRetries)}, {
+							var numFailures = 0;
+							if(testRetry > 0, {
+								"Retry #%... waiting for cooldown".format(testRetry).postln;
+								5.0.wait;
+							});
+							
+							testRetry = testRetry + 1;
+
+							try {
+								result = class.new.runTestMethod(script);
+							} { |error|
+								class.new.failed(script, error.errorString, true, nil);
+							};
+
+							numFailures = IdentitySet.newFrom(class.failures).difference(oldFailures).size;
+							if(numFailures == 0, {testSuccess = true;});
+						});
 
 						endTime = Date.localtime();
 						test[\completed] = true;
@@ -241,3 +257,4 @@ if(File.exists(~testProto).not) {
 	}; // exit only if the script was called with arguments
 }.forkIfNeeded(AppClock);
 )
+

```

**Comment by dyfer** [comment]
Thanks @capital-G 

The implementation looks fine I think...
I don't think we need such a long cooldown, 1 second would probably be enough? Or do you want to play it safe?

----------

From what I observed, we have generally two types of failures in the test suite:
- some tests occasionally failing
- the whole test suite stopping and eventually timing out

This PR would address the first one, hopefully.

The second one _I think_ is caused by a dangling server, i.e. we try to boot the server before the previous instance shuts down and we get "port not available" error. The issue is that we don't have a way to handle boot failure, the test suite simply stops, as we keep waiting for the server.

Also, we don't wait for the server to shut down... I was trying to address that in https://github.com/supercollider/supercollider/pull/6738 (that PR stalled, I wasn't sure if we want to keep it or not, maybe I should re-visit it).
